pulsar-broker/src/main/java/org/apache/pulsar/broker/service/TopicEventsDispatcher.java

```diff
@@ -0,0 +1,136 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.broker.service;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CopyOnWriteArrayList;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * Utility class to dispatch topic events.
+ */
+@Slf4j
+public class TopicEventsDispatcher {
+    private final List<TopicEventsListener> topicEventListeners = new CopyOnWriteArrayList<>();
+
+    /**
+     * Adds listeners, ignores null listeners.
+     * @param listeners
+     */
+    public void addTopicEventListener(TopicEventsListener... listeners) {
+        Objects.requireNonNull(listeners);
+        Arrays.stream(listeners)
+                .filter(x -> x != null)
+                .forEach(topicEventListeners::add);
+    }
+
+    /**
+     * Removes listeners.
+     * @param listeners
+     */
+    public void removeTopicEventListener(TopicEventsListener... listeners) {
+        Objects.requireNonNull(listeners);
+        Arrays.stream(listeners)
+                .filter(x -> x != null)
+                .forEach(topicEventListeners::remove);
+    }
+
+    /**
+     * Dispatches notification to all currently added listeners.
+     * @param topic
+     * @param event
+     * @param stage
+     */
+    public void notify(String topic,
+                       TopicEventsListener.TopicEvent event,
+                       TopicEventsListener.EventStage stage) {
+        notify(topic, event, stage, null);
+    }
+
+    /**
+     * Dispatches notification to all currently added listeners.
+     * @param topic
+     * @param event
+     * @param stage
+     * @param t
+     */
+    public void notify(String topic,
+                       TopicEventsListener.TopicEvent event,
+                       TopicEventsListener.EventStage stage,
+                       Throwable t) {
+        topicEventListeners
+                .forEach(listener -> notify(listener, topic, event, stage, t));
+    }
+
+    /**
+     * Dispatches SUCCESS/FAILURE notification to all currently added listeners on completion of the future.
+     * @param future
+     * @param topic
+     * @param event
+     * @param <T>
+     * @return future of a new completion stage
+     */
+    public <T> CompletableFuture<T> notifyOnCompletion(CompletableFuture<T> future,
+                                                       String topic,
+                                                       TopicEventsListener.TopicEvent event) {
+        return future.whenComplete((r, ex) -> notify(topic,
+                event,
+                ex == null ? TopicEventsListener.EventStage.SUCCESS : TopicEventsListener.EventStage.FAILURE,
+                ex));
+    }
```

Comment:
I see some chained calls like:

```java
                                topicEventsDispatcher.notifyOnCompletion(
                                        topicEventsDispatcher.notifyOnCompletion(res,
                                                topicName.toString(), TopicEvent.CREATE),
                                        topicName.toString(), TopicEvent.LOAD);
```

It looks a little complicated. I also noticed the returned value is never used except this chained call. It's also unnecessary because

```java
future.whenComplete((r, e) -> f1(r,e)).whenComplete((r, e) -> f2(r, e));
```

works nearly the same as

```java
future.whenComplete((r, e) -> { f1(r, e); f2(r,e); });
```

I think a better API could be like:

```java
    public <T> void notifyOnCompletion(CompletableFuture<T> future, String topic,
                                       TopicEventsListener.TopicEvent... events) {
        future.whenComplete((__, ex) -> {
            final var stage = (ex == null)
                    ? TopicEventsListener.EventStage.SUCCESS
                    : TopicEventsListener.EventStage.FAILURE;
            for (var event : events) {
                notify(topic, event, stage, ex);
            }
        });
    }
```

Then we can replace the chained call with:

```java
topicEventsDispatcher.notifyOnCompletion(res, name, TopicEvent.CREATE, TopicEvent.LOAD);
```

Reply:
Good explanation. It makes sense to me.